% !TEX root = ./main.tex
\section{Introduction}
It has been more than sixty years since Jacob and Monod \cite{jacob1961genetic} shaped the way we think about transcriptional regulation in prokaryotes, yet, although about $10^{17}$ bases have been deposited in the SRA database\cite{sragrowth}, we have yet to obtain a full understanding of how all the genes of a single organism are regulated. Even in the case of one of biology's best studied model organism, \textit{Escherichia coli}, about two thirds of the genes lack any regulatory annotation (see \ref{sec:SI_ecoli_ignorance}). For other prokaryotic model organisms the numbers are similar (see \ref{sec:SI_bacillus_ignorance}, \ref{sec:SI_pseudonomas_ignorance}), while higher order model organisms such as \textit{Saccharomyces cerevisiae} (see \ref{sec:SI_yeast_ignorance}) and \textit{C. elegans} (see \ref{sec:SI_elegans_ignorance}) have close to no regulatory annotations, given the arguably more complex nature of gene regulation in eukaryotes. Understanding how genes are regulated is required to understand how an organism adapts its physiology on short time scales to environmental stresses, as well as evolutionary adaption on long time scales. In addition, gene regulation networks and their building blocks, such as transcription factor binding sites and RNA polymerase (RNAP) promoters, are key elements in the design of synthetic gene circuits \cite{elowitz2000synthetic, mangan2003structure, alon2006introduction}.

With its ever increasing availability, Next Gen Sequencing (NGS) is primed to be the method of choice to discover transcription factor and RNAP binding sites. A vast array of methods exists that make it possible to identify binding sites of either specific proteins or for a broad spectrum of DNA binding factors. In methods like ChIP-Seq \cite{rhee2012chip}, proteins have to be cross linked to DNA, which often requires changing residues in the amino-acid sequence, such as for LacI in \textit{E. coli} \cite{rutkauskas2009tetramer}. In addition, antibodies against the protein of interest have to be available, or the protein has to be modified to include a tag which can be targeted by antibodies. While the resolution of these methods is ever improving, it does not allow for a nucleotide resolution yet, making it difficult to identify changes in binding affinity caused by single mutations. Other methods such as ATAC-Seq \cite{buenrostro2015atac, li2019identification} and DNase-Seq \cite{boyle2008high} rely on open chromatin for binding site identification, and have almost exclusively been used for eukaryotic organisms. In general, identifying regulatory interactions from transcription factor occupancy alone can be misleading, since there can be high affinity binding sites in the genome, where there is no change in expression levels upon binding \cite{yona2018random}. DAP-Seq \cite{bartlett2017mapping} is a method similar to ChIP-Seq, however, instead of using immunoprecipitation to obtain DNA-TF pairs, purified and tagged TFs are incubated with fragmented genomic DNA. The method has been used to identify genome wide binding sites for TFs in \textit{Clostridium thermocellum} \cite{hebdon2021genome} and \textit{Riemerella anatipestifer} \cite{zhang2022genome}.

Another approach is to use RNA-Seq as readout for mutagenised promoter regions, where binding sites are identified as regions that, when mutated, lead to significant increase or decrease in expression of a repressor gene \cite{urtecho2018systematic, urtecho2020genome, ireland2020deciphering}. Here we studied the regulatory architecture of 104 genes, including energy matrices with nucleotide resolution that make it possible to build thermodynamic models to predict gene expression \cite{kinney2010using,belliveau2018systematic,barnes2019mapping,ireland2020deciphering}. Additionally, we present major improvements to the method called Reg-Seq \cite{ireland2020deciphering}, making further steps towards obtaining a method allowing to discover regulatory architectures genome wide. Reporter genes are chromosomally integrated into the \textit{E. coli} genome, and reduced diversity in mRNA stability lead to more precise identification of binding sites. A vast array of growth conditions is used to show how certain binding sites can only be identified in a certain growth condition, such as \tr{name example}. The identification of transcription factors was moved on from laborious mass spectrometry experiments, using \textit{in vitro} binding assays as well as a library of transcription factor knockout strains. Finally, improved computational analysis increases the speed of data analysis and the accuracy of parameters that are used for thermodynamic models \tr{here I am thinking Rosalinds stuff}.



