% !TEX root = ./main.tex
\section{Results}
\subsection{Genes studied}
\tr{Does this belong into results or introduction?}
In total we present the regulatory architecture of x promoters, which tells us how a total of y genes are regulated. 18 promoters were chosen as so called "gold standards". These genes have well annotated promoters and have been studied in detail in previous experiments \cite{belliveau2018systematic,ireland2020deciphering}. Including this set of genes allows us to compare the method presented in this work to previous iterations and verify the results, as well as find possible derivations or improvements. x promoters were chosen for genes that have been identified to have a high variation in protein copy number \tr{Probably should show that in the SI} across a set of 22 growth conditions by Schmidt et al., 2016 \cite{schmidt2016quantitative}. These genes were chosen since a high variation in copy number suggests that there are regulatory proteins controlling the expression of the gene. From the same dataset, a set of x promoters was chosen for genes with unidentified function, as annotated by the Schmidt et al., 2016 \cite{schmidt2016quantitative}. None of these promoters had any regulatory annotation prior to the experiments. Another set of x promoters were chosen for genes that were identified in EcoCyc as not having any functional annotation. Two groups of genes, so called iModulons \cite{lamoureux2021precise} were chosen from the work of Lamoureux et al. 2021, where ca. 800 RNA-Seq datasets were evaluated to find genes that were regulated in a distinct network. \tr{Give details for iModulons and their function in their respective paragraphs?} The \tr{Give summary of all genes at the end of this paragragh.}
\tr{Continue with genes of defined circuits and toxin/antitoxin genes}

\subsection{Barcode Mapping}
For each gene studied here, we designed a library of 1500 mutated promoter variants with an average mutation rate of 0.1 for each promoter that was identified with the gene in EcoCyc. If a gene did not have a promoter identified, we first looked for a possible TSS identified by Urtecho et al., 2018 \cite{urtecho2018systematic}, which then was taken as initial sequence for generating mutated variants. In some cases, no TSS could be identified and we used the model of LaFleur et al., 2022 \cite{lafleur2022automated}, which predicts trancsription start site given a genomic sequence of \textit{E. coli}, to find the site in the intergenic region leading up to the first coding region in the operon the gene was part of that had the highes predicit affinity for $\sigma70$-factors. A total of 178619 sequences were ordered from Twist Biosciences. We added random barcodes to the sequences and cloned them into a plasmid vector for amplification and subsequent cloning steps. Details can be found in Supplementary information sections \ref{sec:ident_tss}, \ref{sec:comp_prom_muta}, \ref{sec:library_cloning} and  \ref{sec:barcode_mapping}. 170192 (>95\%) of sequences were identified in the plasmid library (Figure \ref{fig:num_uniq_prom_var}), and an addional 17114 sequences were found, likely due to errors in the synthesis process of the oligonucleotide library. 


\subsection{Genome Integration of Reporters}




\subsection{Transcription Factor identification}
\subsection{Growth Conditions}
\begin{figure}
    \centering
    \includegraphics{../figures/FIG2.pdf}
    \caption{Method summary}
    \label{fig:method_sum}
\end{figure}
\subsection{Gold Standard genes}
\subsection{Ethanol iModulon}
YgeV has been predicted to be a regulator involved in purine catabolism, leading to the production of allantoin, which can be used as a sole nitrogen source \cite{iwadate2019identification}. There are 16 putative regulatory targets \cite{lamoureux2021precise} for YgeV, including the \textit{xdhABC} operon, which degrades xanthine to uric acid \cite{iwadate2019identification} in the purine catabolism pathway. \textit{E. coli} can survive exposure to low ethanol concentrations up to 5\%, which can even lead to increased DNA synthesis \cite{basu1994effect}, but mostly leads to varies stress responses such as an increased production of ROS. Growth in media supplemented with ethanol induced a change in gene expression for genes regulated by YgeV in a \textDelta\textit{baeR} or \textDelta\textit{cpxR} mutant strain. \tr{Is there a correlation between ethanol response and higher need for nitrogen?}
\subsection{Oxidative stress response iModulon}
The putative transcription factor YmfT regulates 14 out of 23 genes in the e14 prophage and is predicted to respond to oxidative stress \cite{lamoureux2021precise}. Oxidative stress is caused by reactive oxygen species (ROS) such as $\text{H}_2\text{O}_2$, which are highly reactive and damage DNA, the cell wall, proteins \cite{ezraty2017oxidative} etc., however, oxidized amino acids can also lead to confirmational changes in transcription factors, such as OxyR and HypT, which induce DNA binding and subsequent regulation of genes involved in response to oxidative stress  \cite{ezraty2017oxidative}. Hydrogen peroxide is produced endogenously in various pathways in \textit{E. coli} and especially in high amounts when phenylethylamine is used as either carbon or nitrogen source \cite{ravindra2013escherichia}, hence we used used minimal media supplemented with 10 mM 2-phenylethylamine hydrochloride (PEA) as sole carbon source to induce stress responses to $\text{H}_2\text{O}_2$ and therefore oxidative stress. \tr{discuss findings of ymfT modulon and how it relates to oxyR, look at oxyR iModulons and possibly do another run including this gene.}
\subsection{Antitoxin/Antibiotic genes}
\subsection{other y-ome genes}
