% !TEX root = ./main.tex
\section{Promoter footprints}
If a base of a binding site for a regulatory element in a promoter is mutated, the expression of the downstream gene is changed due to differences in binding affinity of the regulatory element \tr{cite Kinney, 2010 and Garcia, 2011; but maybe find some older/more original references}. One can generate so called \textit{footprints}, where the effect of a mutation in the promoter on expression levels can be quantified by various metrics. Here, we explore various ways to compute footprints and explain each method in detail.
\tr{add the footprints from one real dataset to compare}
\subsection{Dataset}
For a given promoter, there are $i = 1,..,n$ promoter variants, where each variant has $m_i$ unique barcodes. Per barcode, there are $c_{\mathrm{dna}}$ counts from genomic DNA sequencing, as well as $c_{\mathrm{rna}}$ counts from RNAseq. DNA sequencing is performed to normalize the RNA sequencing data by the abundance of cells in the culture expressing the reporter from a specific promoter variant.
\begin{table}[]
    \centering
    \begin{tabular}{|l|l|l|}
    \hline
    $c_{\mathrm{dna}}$ & $c_{\mathrm{rna}}$ & sequence         \\ \hline
    10               & 2                & ACGTACGTAC\\ \hline
    1                & 2                & ACGTACGTTC\\ \hline
    3                & 5                & ACGTACGTTC\\ \hline
    4                & 9                & ACGTACGTTC\\ \hline
    3                & 5                & ACGTAAGAAC\\ \hline
    3                & 6                & ACGTAAGAAC\\ \hline
    15               & 12               & GCGTACGTAC\\ \hline
    5                 &3                & GCGTACGTAC\\ \hline
    12               & 14               & ACATACGTAC\\ \hline
    2                & 3                & ACATACGTAC\\ \hline
    20               & 40               & ACATACGTAC\\ \hline
    5                & 3                & ACGGATGTAC\\ \hline
    5                & 1                & ACGTACGTGA\\ \hline
    10               & 1                & ACGTACGTGA\\ \hline
    2                & 10               & ACGTCCATAC\\ \hline
    2                & 10               & ACGTCCATAC\\ \hline
    4                & 13               & ACGTCCGTAC\\ \hline
    18               & 25               & ACAAACGTAC\\ \hline
    17               & 19               & GCGTACGTAG\\ \hline
    10               & 11               & GCGTACGTAG\\ \hline
    2                & 3                & GGGTACGTAG\\ \hline    \end{tabular}
    \caption{Example dataset, arbitrarily generated. Different counts for the same sequence come from unique barcodes, are therefore separate measurements.}
    \label{tab:exaple_data}
\end{table}
\subsection{Expression Shifts}
Belliveau et al. (2018)\cite{belliveau2018systematic} used so called \textit{expression shifts} to compute footprints for mutagenized promoters. In their experiments, cells were sorted based on fluorescence, where the fluorescent reporter gene was expressed under the control of a mutagenized promoter variant, and subsequently sequenced. Therefore, each sequence had a bin associated with it, which is a read out for how strong the reporter is expressed relatively to the other promoter variants in the library. This approach can be adapted to our data set, where we first compute the average relative expression $\langle c \rangle_i$ per promoter variant across all of its unique barcodes,
\begin{equation}
    \langle c \rangle_i = \frac{1}{m_i}\sum_{j=1}^{m_i} \frac{c_{\mathrm{rna}, j}}{c_{\mathrm{dna}, j}}.
\end{equation}
Then, we determine how much relative expression is changed at each position if there is a mutation. If a base at position $\ell$ in promoter variant $i$ is mutated, we denote that as $\sigma_{i, \ell}=1$. Otherwise, if the base is wild type, we write $\sigma_{i, \ell}=0$. Then, the change in relative expression due to mutation, the expression shift $\Delta c_\ell$, at position $\ell$ is given by
\begin{equation}
   \Delta c_\ell =\frac{1}{n}\sum_{i=1}^n \sigma_{i, \ell} \left( \langle c \rangle_i - \frac{1}{n}\sum_{k=1}^n \langle c \rangle_i \right).
\end{equation}
\begin{figure}
    \includegraphics[scale=1]{../figures/example_expression_shift.pdf}
    \caption{Expression shift for example dataset. \tr{Add expression shift for real dataset.}}
    \label{fig:SI_expression_shift}
\end{figure}
The absolute value of expression shift can be hard to interpret, so indeed one can present it in terms of relative change to the mean expression, i.e., fold-change,
\begin{eqnarray}
    \delta c_\ell = \frac{\Delta c_\ell}{\langle c \rangle} =\frac{1}{n}\sum_{i=1}^n \sigma_{i, \ell} \left( \frac{\langle c \rangle_i}{\langle c \rangle} - 1 \right).
\end{eqnarray}
Figure~\ref{fig:SI_expression_shift} shows the expression shift footprint that is obtained for the test dataset. \tr{as well as the footprint for a real data set}.
\subsection{Frequency Matrices}
\tr{Not sure if I will actually write about it, just a different way of computing footprints I came up with based on comments by Frank J\"ulicher and Stephan Grill. Have try it on old data set.}
\subsection{Mutual Information}
Mutual information is a measure of much information is obtained about a random variable by measuring a different random variable. In the context of gene expression, this can be understand as the ability to predict changes in gene expression given a certain mutation on the promoter sequence. If there is no annotation, meaning it us unknown where RNAP or transcription factors bind, one can not make any predictions on the expression level of the downstream gene when observing a mutation in the promoter. In this case, there is low mutual information between sequence and expression level. On the other hand, if the promoter is annotated and on has binding energy matrices for all transcription factor binding sites and the RNAP binding site in hand, then one can precisely predict the change in gene expression given any point mutation based on thermodynamic models \tr{could cite a bunch of papers here}, which is a case of high mutual information. Hence, by maximizing the mutual information between a model for the regulatory architecture and observed levels of gene expression, we can discover binding sites for transcription factors and subsequently, using equilibrium thermodynamic models and neural networks, compute binding energy matrices in real units of $k_BT$.

\subsubsection{Mutual Information based on Sequence Counts}
The first way of computing mutual information at each position in the promoter is to take the base at each position as one random variable, and the expression of each sequence as other random variable. As measure for expression, we use RNA counts for each sequence normalized by DNA counts. In order to compute mutual information, we need to obtain a probability distribution $p_\ell(c, \mu)$, which gives the probability of finding a certain base $c$ at position $\ell$, and corresponding expression $\mu$. One way of obtaining such a distribution is to find bins for the values of $\mu$, denoted as $\mu_b$, as shown in Figure~\ref{fig:SI_MI_exp_bins}. Then, mutual information is given by
\begin{equation}
    I_\ell = \sum_{c=\mathrm{A,C,G,T}} \sum_{\mu_b}p_{\ell}(c, \mu_b)\log_2\left(\frac{p_{\ell}(m, \mu_b)}{p_{\ell}(c)p(\mu_b)}\right),
\end{equation}
where $p_{\ell}(c)$ and $p(\mu_b)$ are the marginal distributions.





\iffalse
The most direct way of computing mutual information at each position in the promoter is to take the base identity, i.e. wild type or mutation, as one random variable, and the sequencing counts as other random variable. Mutual information $I_\ell$ at position $\ell$ is then given by
\begin{equation}
    I_\ell = \sum_{\mu=0, 1}\sum_{m=0, 1} p_{\ell}(m, \mu)\log_2\left(\frac{p_{\ell}(m, \mu)}{p_{\ell}(m)p_(\mu)}\right),
\end{equation}
where $m=0$ denotes the wild type base, $m=1$ denotes a mutated base, $p(m)$ is the fraction of either mutated or wild type bases in the sequencing data, $\mu=0$ denotes a DNA read, $\mu=1$ denotes a RNA read, $p(\mu)$ is the fraction of either DNA or RNA counts in the sequencing data and $p(m, \mu)$ is the fraction of either wild type or mutation in either DNA or RNA reads. The index $\ell$ denotes that these fractions are computed at each position independently. For the test dataset in table \ref{tab:exaple_data} we can compute these fractions. There are a total of 181 sequencing counts, 88 of which belong to reads from DNA, therefore, $p(\mu=0) = 88/181$ and $p(\mu=1)=93/181$, since the rest of the reads are from RNA.
\begin{equation}
    p(\mu) = \begin{cases}
        88/181\quad\mu=0,\\
        93/181\quad\mu=1.
    \end{cases}
\end{equation}
At position 1, row 4 and 10 have a mutation, therefore, 32 reads are from DNA with a mutation, and 31 reads are from RNA with a mutation. Hence, $p_{1}(m, \mu)$ is given by 
\begin{equation}
    p_1(m, \mu) = \begin{cases}
        56/181\quad m=0,\mu=0,\\
        62/181\quad m=0,\mu=1,\\
        32/181\quad m=1,\mu=0,\\
        31/181\quad m=1,\mu=1.
    \end{cases}
\end{equation}
The fraction of reads with a mutation across both DNA and RNA reads is then simply given by taking the sum across $\mu$,
\begin{equation}
    p_1(m) = \begin{cases}
        118/181\quad m=0,\\
        63/181\quad m=1.
    \end{cases}
\end{equation}
Having these values in hand, we can compute mutual information and the first base. The mutual information footprint based on base identity for the example dataset is shown in Figure~\ref{fig:SI_MI_base_ident}. There is a clear distinction between the footprint obtained from expression shift compared to this way of computing mutual information. \tr{add some explanation on why that is, compare to eLife paper.}
\fi
\begin{figure}
    \includegraphics[scale=1]{../figures/example_MI_exp_bins.pdf}
    \caption{Possible binning of expression counts for example data set. \tr{Add footprint for real dataset.}}
    \label{fig:SI_MI_exp_bins}
\end{figure}
\begin{figure}
    \includegraphics[scale=1]{../figures/example_MI_bases_bins.pdf}
    \caption{Mutual information based on base identity for example dataset. \tr{Add footprint for real dataset.}}
    \label{fig:SI_MI_base_ident}
\end{figure}
\subsubsection{Mutual Information based on Phenotype Matrices}
A different way to utilize mutual information is to choose a phenotype as random variable instead of base identity. The phenotype can be computed from the sequence using e.g. an additive model, where each position independently contributes to the total value of the phenotype $\Phi$,
\begin{equation}
    \Phi = \Theta_0 + \sum_{l=1}^{L}\sum_c \Theta_{l:c}x_{l:c},
\end{equation}
where $\Theta_{l:c}$ is the phenotype matrix, $\Theta_0$ an additive constant and $x_{l:c}$ is a one-hot representation of the sequence with
\begin{equation}
    x_{l:c}=\begin{cases}
        1\quad\text{if character c occurs at position l},\\
        0\quad\text{otherwise}
    \end{cases}
\end{equation}
where the notation is adapted from \cite{tareen2022mave}. Without any knowledge of the regulatory architecture of the promoter, one can only make random guesses for the phenotype matrix. However, either using Metropolis-Hasting algorithms \tr{Reg-Seq, gotta decide how much to write about it} or Neural Networks \tr{MaveNN, will be included if we get good results with it}, the phenotype matrix can be optimized in the sense its entries are more extreme where there are binding sites for regulatory elements in the sequence, since a mutation in that part of the sequence will have the strongest effect on gene expression. How extreme entries are can be quantified by using relative entropy, where the entries for each position on the sequence are first converted to a probability distribution using exponential weights, and then Kullback-Leiback-Divergence (KLD) between the resulting distribution and a uniform distribution is calculated. 
\tr{Expand by explaining how peaks are identified as binding sites.}
 
\subsubsection{Phenotype Matrices and Neural Networks, MaveNN}
\subsubsection{Identifying Binding Energy Matrices}